### DISCUSSION TITLE
Can't capture video -- unusual behaviour in Firefox

### DISCUSSION THREAD

**Discussion by computerologist** [aclap-dev/video-downloadhelper#219] [Help and Q&A]
### Your question

In firefox, I noticed that I couldn't get a screen capture of this video when doing a screenshot:

https://watch.animationshowcase.com/65d6c14cef9b464a320712db/assets/war-is-over-private-access?autoPlay=true&enableBack=true

It just comes out black.

I also get nothing when I use OBS.

This isn't specific to downloadhelper, but the video also doesn't show up in downloadhelper.

I can take a screenshot normally in chrome.

Maybe someone in here will have some insight.  I've never seen behaviour like this before.

### Addon & CoApp version and details

version: 8.1.4.0
beta: false
build date: 2024-01-04
build options: linuxlic: false, noyt: false, browser: firefox
lang: en-GB
license: unset
platform: x86-64 win
UA: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:123.0) Gecko/20100101 Firefox/123.0
coapp: yes. Info:
{
  "id": "net.downloadhelper.coapp",
  "name": "VdhCoApp",
  "version": "2.0.10",
  "binary": "C:\\Program Files\\DownloadHelper CoApp\\vdhcoapp.exe",
  "displayName": "VdhCoApp",
  "description": "Video DownloadHelper companion app",
  "target": {
    "os": "windows",
    "arch": "x86_64",
    "node": "18"
  },
  "home": "REDACTED",
  "converterBinary": "C:\\Program Files\\DownloadHelper CoApp\\ffmpeg.exe",
  "converterBase": "ffmpeg",
  "converterBaseVersion": "6.0"
}

**Comment by Mattmjs** [comment]
DRM protected and there is no way around that.

**Reply by computerologist** [reply]
I mean obviously there's a way around it since it's leaving my video card and hitting my monitor.

**Reply by Mattmjs** [reply]
Well of course, the video player decrypts the DRM protection so that you can view it on the web, but if you attempt to
download it you end up with a scrambled encrypted file. I did that and that's what I got.
Which means downloading is not an option.

**Reply by computerologist** [reply]
I guess this post isn't specific to vdh really.  Just casting about trying to understand what's happening here.  I haven't seen anything that resisted being screenshotted before, and it's odd that it's browser specific.

**Reply by Mattmjs** [reply]
Do you mean you can't take a screenshot or do you want to download it ? I would have thought it would work in OBS
if you want to screen record it.

**Reply by GrampaWildWilly** [reply]
It refuses to take a screenshot.  I tried with my usual PrintScreen key & I tried taking a screen capture using Fast Stone Image Viewer.  All I got was a black image.  I've never seen that before, either.  I haven't installed OBS yet on this system (W11, using only a couple of months) so I can't say what it does.  But failure to take a screenshot seems awfully weird.  This site must be doing something extremely odd.

Wait.  I tried using Firefox's own built-in screenshot taker.  Pop up a context menu anywhere on the screen, then select Take Screenshot, then select Save Full Page.

![#01](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/6b1e0fa4-1039-407c-bc6f-ac54b19d75f0)

This is all quite strange.  It doesn't get you any closer to downloading this, which Matt explained above.  But at least it took a screenshot.

**Reply by computerologist** [reply]
@GrampaWildWilly I don't really need to download this particular video, just more interested in what's going on here, since I have not seen anything like this before. I hope it doesn't spread though...but if it's effective it will. I only noticed it because I tried screenshotting something.

Yeah I tried the windows native snip and sketch: black image. OBS: black image. Picpick: black image.

But yeah the firefox native one gives a screenshot--and I can use the apps above if it's running in Chrome. So strange.

@Mattmjs I'd like to understand how it's even possible for a website to prevent a screenshot being taken.  Even if I have focus elsewhere, the video is running, but if I enter the SS context, it goes black.  I assume that this is the tip of an iceberg we are going to start seeing more, as this is new behaviour.  So yes being able to download it entirely would be a nice thing to have too, but that's probably much harder than getting a screenshot.

**Reply by Mattmjs** [reply]
I see whenever the video is paused it goes dark, so it is just something they've done the way it is programmed or coded.
But there is a workaround , use the video pop out feature built into Firefox, other browsers might have it also.
Get the frame you want then pause it and use printscreen button.

**Reply by computerologist** [reply]
@Mattmjs I thought it might be that at first, but it's not.  If you let it play and do a screencap while it's playing, the play area will be black--same with OBS.  As @GrampaWildWilly said above you can use the firefox native capture to work around it...but still doesn't really answer the question of what's going on.

Yes it's definitely probably deliberate on their part.

**Reply by GrampaWildWilly** [reply]
Oh, that's a thought.  Play it in picture-in-picture.  Nope.  Still can't take a standard screenshot.  The Firefox context menu doesn't work in the picture-in-picture window.  It just doesn't pop up.  While the video is playing in picture-in-picture, you can still pop up the Firefox context menu . . . on the main window, where it says it's playing in picture-in-picture mode.  But what it captures is that main window, which contains just the message that it's playing in picture-in-picture.  Fat lot of good that is.  What we want to capture is, as it says, playing in the picture-in-picture window, & the screenshot doesn't capture that.  And the usual screenshot techniques, PrintScreen or your image processing application's screen capture function, still cause the picture-in-picture window to go temporarily dark when you try those.  I don't know what the deal is with this.  And you say OBS fails, too?  I can't explain it.  We don't appear to be talking about VDH any more so this has gotten way off topic.  We established rather a while ago that this particular video is behind DRM protection so VDH is doomed to fail on this.  It is odd that OBS is also failing, but that is a topic for another forum.  Does this presage a trend?  I wouldn't jump to that conclusion just yet.  One data point does not constitute a trend.  But we might want to stay vigilant for such a trend.

**Reply by computerologist** [reply]
Yeah I don't want to clutter this forum with stuff that isn't directly related.  Just unsure who might be interested in talking about it.  Maybe no one, if this is the only example.

**Reply by GrampaWildWilly** [reply]
I don't think we've made it clear here yet.  Neither Matt nor I works for VDH.  We are both just users, like most people posting in this forum.

**Reply by GrampaWildWilly** [reply]
One last idea Matt proposed.  If you pause playback in the picture-in-picture window, the paused frame is visible.  But none of the screenshot techniques works, not PrintScreen, not image processor screen capture, not the Firefox popup menu screenshot.  This is one resistant little bugger.

**Reply by computerologist** [reply]
Oh I think a dev would have shut this down already lol.  But it's so weird that I don't get any of this obfuscation in Chrome.  I wonder if it's not actually deliberate.

**Reply by computerologist** [reply]
Just tried in Edge as well.  Same behaviour as firefox, down to being able to use the native region capture but nothing else.  Weird because Edge is Chromium, like Chrome.

**Reply by GrampaWildWilly** [reply]
Well, after all this off-topic natter, why not some more?  I installed OBS.  I've just been in the OBS support chat on Discord.  What a horrific interface to get help.  The search is unintelligible.  Anyway, some kind person tells me that OBS does not record DRM-protected content.  I could swear I managed to record some DRM content when I was using a rather older version of OBS on Windows 7.  That was at least a year ago.  This new one does have some improvements on features that didn't work properly on W7.  It seems to record XM radio OK, although VDH & ffmpeg aren't any help with that.  But it seems OBS have succumbed to the pressure & they no longer record DRM content.  Such a shame.  I shall have to stop recommending screen recorders for recording DRM content.  Unless some kind person here can name another screen recorder that isn't deterred by DRM.  I may investigate this further myself.  I thought this was a battle that was won when VCRs became commonplace on the consumer market.  That was what?  50 years ago?  A screen recorder is no different than a VCR, on a certain level.  It should be possible to record anything that appears on your monitor.  I am not a happy camper.

**Reply by GrampaWildWilly** [reply]
I actually installed ScreenRec & Bandicam.  Neither one recorded my livestream from my cable provider.  They have limits on the duration they will record & they add a watermark if you don't pay extra.  I didn't investigate their prices since their free versions didn't work anyway.  I think FreeRec saved to the cloud, which is unacceptable to me.  In addition, the files of what they installed had a lot of file names beginning with QT.  I think that means it's QuickTime based.  Not too impressed with that.  Bandicam did have a nicer UI, but that's not significant when the thing doesn't work.  I Revo Uninstalled them after my tests.  I also found something called TinyTake, but they don't have a free download, just an exhortation to schedule a free demo.  If anybody else finds anything, do post.  I'm abandoning my search.

I tested OBS against the livestream from my cable TV provider, NBC Sports, & Tennis Channel.  OBS detects a black screen on all of those.  I went to my favorite hockey site, which at this time of day is not showing NHL games because there aren't any in progress.  But it was showing other leagues.  I picked one at random & it could see that.  Plus it works on XM.  So it's not a complete bust.  But copyright protected material appears to now be a problem.  At some point within the past year, all the heavyweights appear to have colluded to prevent screen recorders from working.  I say again, OBS used to work on Golf Channel's livestream from NBC Sports & my cable provider's livestream.  I used it a couple of times back when I was on Windows 7.  I also OBSed a 2-day rental thing I bought from YouTube.  I think it was DRM protected but I didn't positively verify that at the time.  VDH downloaded one of those garbled things it does on DRM content so I'm guessing it was behind DRM protection.  But that was a year ago & more.  Things have changed.  Changed for the worse.

**Reply by computerologist** [reply]
The arms race continues.  I too hate the idea that a machine I own is fundamentally crippled by third party software running on a website.  There must be a way to capture the framebuffer right before it goes to a monitor.

DRM software in the past wasn't able to interfere with the basic function of a machine, which is how I see this.  I suppose eventually one will need to connect the video out to another device in between the VC and the monitor.

**Reply by GrampaWildWilly** [reply]
I don't think it's so much DRM software interfering with the machine.  I think it's the software builders all being coerced -- lawyers are involved -- to make their software recognize when something has DRM on it & decide to fail to process it the way they should, they way they used to, the way they easily could if they weren't being threatened.

Just as an aside, I did continue my search.  I found several more screen recorders that all interpreted my cable system's livestream as a black image.  I did find one screen recorder (iTop Screen Recorder) that did record my TV feed.  But there were so many negatives with it that I couldn't deal with it.  First, it required so much fiddling to get it to do the recording that I couldn't reliably make the thing work.  It also did extremely strange things to my system, among them permanently hiding my Task Bar, making it impossible to switch tasks, making it impossible to make the application itself reappear which meant I couldn't stop it from recording, changing the sizes & positions of unrelated open windows, swapping the relative positions of my monitors, swapping unrelated windows from one monitor to the other, and more.  What finally brought some order back to proceedings was that it stopped recording after 10 minutes.  That's more of what's wrong with the thing.  It has a limit of 10 minutes on what you can record & it places the name of the software vendor in the middle of the recorded image as a watermark.  Note that word: vendor.  They will sell you the ability to record for longer without a watermark.  I didn't even try to find out what it cost.  Once it finally let go of my system, I Revo Uninstalled it.  As a parting shot, it told me that the uninstall wouldn't complete until I reboot, which I have not done yet because I am listening to tonight's live Metropolitan Opera broadcast, which is being streamed live on the Met's web site.  No.  Wait.  That was another screen recorder, whose name I have wiped from my memory, that did that.  It can wait until after the opera before I finish removing it.  I already cleaned up after iTop by rebooting.  Now I really am giving up.
